server/src/domain/person/person.service.ts

```diff
@@ -442,46 +443,59 @@ export class PersonService {
     const mergeIds = dto.ids;
+    if (!(mergeIds instanceof Array)) {
+      throw new BadRequestException('ids is not an array');
+    }
     await this.access.requirePermission(auth, Permission.PERSON_WRITE, id);
-    const primaryPerson = await this.findOrFail(id);
+    let primaryPerson = await this.findOrFail(id);
     const primaryName = primaryPerson.name || primaryPerson.id;
 
     const results: BulkIdResponseDto[] = [];
 
     const allowedIds = await this.access.checkAccess(auth, Permission.PERSON_MERGE, mergeIds);
 
-    for (const mergeId of mergeIds) {
-      const hasAccess = allowedIds.has(mergeId);
+    for (let i = 0; i < mergeIds.length; i++) {
+      const hasAccess = allowedIds.has(mergeIds[i]);
       if (!hasAccess) {
-        results.push({ id: mergeId, success: false, error: BulkIdErrorReason.NO_PERMISSION });
+        results.push({ id: mergeIds[i], success: false, error: BulkIdErrorReason.NO_PERMISSION });
         continue;
       }
 
       try {
-        const mergePerson = await this.repository.getById(mergeId);
+        const mergePerson = await this.repository.getById(mergeIds[i]);
         if (!mergePerson) {
-          results.push({ id: mergeId, success: false, error: BulkIdErrorReason.NOT_FOUND });
+          results.push({ id: mergeIds[i], success: false, error: BulkIdErrorReason.NOT_FOUND });
           continue;
         }
-
+        if (mergeIds.length === 1 && i === 0) {
+          if ((!primaryPerson.name && mergePerson.name) || (!primaryPerson.birthDate && mergePerson.birthDate)) {
+            primaryPerson = await this.repository.update({
+              id: primaryPerson.id,
+              name: mergePerson.name || undefined,
+              birthDate: mergePerson.birthDate || undefined,
+            });
+          }
+        }
         const mergeName = mergePerson.name || mergePerson.id;
-        const mergeData: UpdateFacesData = { oldPersonId: mergeId, newPersonId: id };
+        const mergeData: UpdateFacesData = { oldPersonId: mergeIds[i], newPersonId: id };
         this.logger.log(`Merging ${mergeName} into ${primaryName}`);
 
         await this.repository.reassignFaces(mergeData);
         await this.jobRepository.queue({ name: JobName.PERSON_DELETE, data: { id: mergePerson.id } });
 
         this.logger.log(`Merged ${mergeName} into ${primaryName}`);
-        results.push({ id: mergeId, success: true });
+        results.push({ id: mergeIds[i], success: true });
       } catch (error: Error | any) {
-        this.logger.error(`Unable to merge ${mergeId} into ${id}: ${error}`, error?.stack);
-        results.push({ id: mergeId, success: false, error: BulkIdErrorReason.UNKNOWN });
+        this.logger.error(`Unable to merge ${mergeIds[i]} into ${id}: ${error}`, error?.stack);
+        results.push({ id: mergeIds[i], success: false, error: BulkIdErrorReason.UNKNOWN });
       }
     }
-
-    return results;
+    return {
+      results,
+      person: primaryPerson,
+    };
```

Comment:
I don't like this return structure very much. It makes me think this endpoint is doing too many things. I'd prefer to either:
1. Leave the code to update the person and then have the web assume the person changed and call get person by id after the merge completes, or 
2. Make two separate requests from the web: one to merge people and a second one to update the person

Basically, this looks bad to me and it would be better to split up the requests to be merge + refresh or merge + update

Reply:
You're right. let's change it; I like option 1